lib/sycamore/sycamore/connectors/neo4j/neo4j_writer.py

```diff
@@ -0,0 +1,343 @@
+                        headers1["nodes"][key][value] = True
+            #### merge relationships together ####
+            for key, values in headers2["relationships"].items():
+                if headers1["relationships"].get(key, None) is None:
+                    headers1["relationships"][key] = values
+                else:
+                    for value in values:
+                        headers1["relationships"][key][value] = True
+            return headers1
+
+        def finalize(nodes):
+            return nodes
+
+        aggregation = AggregateFn(
+            init=lambda group_key: dict(nodes=dict(), relationships=dict()),
+            accumulate_row=accumulate_row,
+            merge=merge,
+            finalize=finalize,
+            name="output",
+        )
+
+        return self._dataset.aggregate(aggregation)["output"]
+
+    def _write_nodes_csv_headers(self, nodes):
+        for node_label, node_columns in nodes.items():
+            csv_path = os.path.expanduser(f"{self._import_dir}/sycamore/nodes/{node_label}.csv")
+            headers = sorted(set(column for column in node_columns.keys()))
+            os.makedirs(os.path.dirname(csv_path), exist_ok=True)
+            with open(csv_path, "w", newline="") as file:
+                writer = csv.writer(file)
+                writer.writerow(headers)
+
+    def _write_relationships_csv_headers(self, relationships):
+        for relationship_label, relationship_columns in relationships.items():
+            csv_path = os.path.expanduser(f"{self._import_dir}/sycamore/relationships/{relationship_label}.csv")
+            headers = sorted(set(column for column in relationship_columns.keys()))
+            os.makedirs(os.path.dirname(csv_path), exist_ok=True)
+            with open(csv_path, "w", newline="") as file:
+                writer = csv.writer(file)
+                writer.writerow(headers)
+
+
+class Neo4jWriteCSV(MapBatch, Write):
+    def __init__(self, plan: Node, client_params: Neo4jWriterClientParams, **kwargs):
+        self._import_dir = client_params.import_dir
+        self._name = "Neo4jWriter"
+        super().__init__(plan, f=self._write_docs_tt, **kwargs)
+
+    def _parse_docs(self, docs):
+        include = ["type", "bbox", "text_representation"]
+        nodes = defaultdict(list)
+        relationships = defaultdict(lambda: defaultdict(list))
+        for doc in docs:
+            if "label" not in doc.data or "doc_id" not in doc.data:
+                continue
+            node = {
+                "uuid:ID": doc.doc_id,
+            }
+            if doc["type"] == "table":
+                doc["text_representation"] = doc["table"].to_csv()
+            for key, value in doc.data.items():
+                # add flatten properties to node
+                if key == "properties":
+                    if isinstance(value, dict) and value:
+                        properties_flat = pd.json_normalize(value, sep="_").to_dict(orient="records")[0]
```

Comment:
There's a fn in sycamore.connectors.common called flatten_data (or something like that) that does this without sticking it in a pandas DataFrame first

Reply:
fixed, no longer use pandas